src/ops/jobset_wait.rs

```diff
@@ -0,0 +1,109 @@
+use crate::ops::{OpResult, OpError};
+use crate::hydra::JobsetOverview;
+use crate::query::jobset_overview;
+use std::time::SystemTime;
+use std::time::{Duration, UNIX_EPOCH};
+use std::{thread};
+use std::option::Option;
+use std::io;
+use std::io::Write;
+
+fn evaluation_started_since(jobset: &JobsetOverview) -> Option<Duration> {
+    let starttime = jobset.starttime?;
+    match SystemTime::now().duration_since(UNIX_EPOCH + Duration::from_secs(starttime)) {
+        Ok(n) => Some(n),
+        Err(_) => None
+    }
+}
+
+fn jobset_find(host: &str, project_name: &str, jobset_name: &str) -> Result<JobsetOverview, OpError> {
+    let jobsets = jobset_overview(host, project_name)?;
+    jobsets
+        .into_iter()
+        .find(|j| j.name == jobset_name)
+        .ok_or_else(|| OpError::Error(
+            format!("Project {} doesn't have a jobset {}",
+                    project_name,
+                    jobset_name)))
+}
+
+// To know if a jobset has been successfully built, four steps are required:
+// 1. ensure a evaluation is not running
+// 2. wait for a new evaluation to start
+// 3. wait for this evaluation to be terminated
+// 4. wait for all scheduled builds to terminate
+//
+// If a build or the evaluation fail, it immediately returns an error.
+//
+// There are several improvements, such as
+// - use the checkinterval to know when the next evaluation will start
+// - use the push Hydra API to trigger an evaluation (but this needs credentials)
+pub fn run(host: &str, project_name: &str, jobset_name: &str) -> OpResult {
+    enum State {
+        WaitingForPreviousEval,
+        WaitingForNewEval,
+        Evaluating,
+        Building,
+    };
+    let sleep = 2;
+    let mut state = State::WaitingForPreviousEval;
+    let mut start = SystemTime::now();
+    let mut nrscheduled = 0;
+
+    println!("waiting for a potential evaluation to terminate");
+    loop {
```

Comment:
The fact that the state machine implementation and actual logic is not separated in any way makes it a bit hard to follow for me. I haven't completely read it but have you maybe had a look at https://hoverbear.org/2016/10/12/rust-state-machine-pattern/ which goes into detail to describe how to model state machines in Rust with different suggestions. It would be very nice if it was easier to see and codify what transitions exists and ideally reflect it on the type system. At the moment it is very imperative and implicit.

Other than that:  maybe you could extract the state transition functions?

Reply:
I simplified the code by creating more functions to get the status of the jobset. I however didn't change the implementation since I don't want to make it more complicated... or I just didn't figure out how to make it more simple!